### PR TITLE
scylla-jmx listen address

### LINKED ISSUE #390: The assigned jmx-port might be in use for large clusters
Seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/37/testReport/update_cluster_layout_tests/TestLargeScaleCluster/Run_Dtest_Parallel_Cloud_Machines___LongDtest___long_split000___test_add_50_nodes/
```
ccmlib.node.NodeError: Error starting node node32: unable to connect to scylla-jmx port 127.0.21.32:10221
```

https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-release/37/artifact/logs-long.release.000/1655881382174_update_cluster_layout_tests.py%3A%3ATestLargeScaleCluster%3A%3Atest_add_50_nodes/node32_system.log.jmx
```
Using config file: /jenkins/workspace/scylla-master/dtest-daily-release/scylla/.dtest/dtest-97fehkcv/test/node32/conf/scylla.yaml
Error: JMX connector server communication error: service:jmx:rmi://47a05c919e49:10221
sun.management.AgentConfigurationError: java.rmi.server.ExportException: Port already in use: 10221; nested exception is: 
	java.net.BindException: Address already in use (Bind failed)
	at sun.management.jmxremote.ConnectorBootstrap.exportMBeanServer(ConnectorBootstrap.java:800)
	at sun.management.jmxremote.ConnectorBootstrap.startRemoteConnectorServer(ConnectorBootstrap.java:468)
	at sun.management.Agent.startAgent(Agent.java:262)
	at sun.management.Agent.startAgent(Agent.java:452)
Caused by: java.rmi.server.ExportException: Port already in use: 10221; nested exception is: 
	java.net.BindException: Address already in use (Bind failed)
	at sun.rmi.transport.tcp.TCPTransport.listen(TCPTransport.java:346)
	at sun.rmi.transport.tcp.TCPTransport.exportObject(TCPTransport.java:254)
	at sun.rmi.transport.tcp.TCPEndpoint.exportObject(TCPEndpoint.java:412)
	at sun.rmi.transport.LiveRef.exportObject(LiveRef.java:147)
	at sun.rmi.server.UnicastServerRef.exportObject(UnicastServerRef.java:237)
	at sun.management.jmxremote.ConnectorBootstrap$PermanentExporter.exportObject(ConnectorBootstrap.java:199)
	at javax.management.remote.rmi.RMIJRMPServerImpl.export(RMIJRMPServerImpl.java:146)
	at javax.management.remote.rmi.RMIJRMPServerImpl.export(RMIJRMPServerImpl.java:122)
	at javax.management.remote.rmi.RMIConnectorServer.start(RMIConnectorServer.java:404)
	at sun.management.jmxremote.ConnectorBootstrap.exportMBeanServer(ConnectorBootstrap.java:796)
	... 3 more
Caused by: java.net.BindException: Address already in use (Bind failed)
	at java.net.PlainSocketImpl.socketBind(Native Method)
	at java.net.AbstractPlainSocketImpl.bind(AbstractPlainSocketImpl.java:387)
	at java.net.ServerSocket.bind(ServerSocket.java:390)
	at java.net.ServerSocket.<init>(ServerSocket.java:252)
	at java.net.ServerSocket.<init>(ServerSocket.java:143)
	at sun.rmi.transport.proxy.RMIDirectSocketFactory.createServerSocket(RMIDirectSocketFactory.java:45)
	at sun.rmi.transport.proxy.RMIMasterSocketFactory.createServerSocket(RMIMasterSocketFactory.java:345)
	at sun.rmi.transport.tcp.TCPEndpoint.newServerSocket(TCPEndpoint.java:670)
	at sun.rmi.transport.tcp.TCPTransport.listen(TCPTransport.java:335)
	... 12 more
```

The jmx port is calculated in https://github.com/scylladb/scylla-ccm/blob/c2f6c70efbbe62c55a108597113bd12eb53cabfd/ccmlib/cluster.py#L329-L330

So for large enough clusters we get arbitrary high ports, like 10221 above (7000 + 32*100 + 21).

### PR DESCRIPTION
**Pull request by bhalevy**
This series starts scylla-jmx with the node address and a pre-defined port (7199)
to prevent clashes with in-use ports.

Fixes #390

### CODE REVIEW

File: ccmlib/scylla_node.py [hunk 1]
```diff
@@ -180,7 +180,9 @@ def _start_jmx(self, data):
         args = [jmx_java_bin,
                 '-Dapiaddress=%s' % data['listen_address'],
                 '-Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder',
+                '-Djava.rmi.server.hostname=%s' % data['listen_address'],
```
Comment (fruch) [inline ccmlib/scylla_node.py:183]: Sure we need both configurations ?

File: ccmlib/scylla_node.py [hunk 2]
```diff
@@ -194,7 +196,10 @@ def _start_jmx(self, data):
         env_copy = os.environ
         env_copy['SCYLLA_HOME'] = self.get_path()
 
+        message = f"Starting scylla-jmx: args={args}"
+        self.debug(message)
```
Comment (fruch) [inline ccmlib/scylla_node.py:200]: I think debug print in one place should be enough, wouldn't it ?

File: ccmlib/scylla_node.py (hunk 1 above)
Comment (bhalevy) [reply ccmlib/scylla_node.py:183]: I didn't try without this one.
The other way around didn't work
and it seems weird to provide the host address one way and the port in another way.

File: ccmlib/scylla_node.py (hunk 2 above)
Comment (bhalevy) [reply ccmlib/scylla_node.py:200]: For scylla we have a debug log that goes to the test log and scylla itself echoes its args to its system log.

I find it useful to have it in both places...

File: ccmlib/scylla_node.py (hunk 1 above)
Comment (bhalevy) [reply ccmlib/scylla_node.py:183]: Yeah, both are needed.
Testing with `pytest -s -vv --cassandra-dir=$CASSANDRA_DIR -n 8 nodetool_additional_test.py`

**Comment by bhalevy** [comment]
Please don't merge yet since I'm seeing failures with nodetool_additional_test.py

**Comment by bhalevy** [comment]
The failures look like the result of running multiple test cases concurrently that use cassandra-stress, and unrelated to this change.  I'm double checking that.

File: ccmlib/scylla_node.py (hunk 2 above)
Comment (fruch) [reply ccmlib/scylla_node.py:200]: o.k.

File: ccmlib/scylla_node.py (hunk 1 above)
Comment (fruch) [reply ccmlib/scylla_node.py:183]: o.k., today won't be the day I understand java configuration logic.

**Comment by bhalevy** [comment]
In v2 (aab480c9d06236b9b94503664afed27f33bc47cb):
- Changed default port to 7199, same as Cassandra's

With additional ignore of expected_errors the test_concurrent_* dtests pass as well.
Will send a PR for ignoring those.

**Review by fruch — APPROVED** [review]
LGTM

**Comment by avelanarius** [comment]
This broke Java Driver’s integration tests.
```
Error:  Failures: 
Error:    CCMBridgeTest.should_make_JMX_connection:46 » IO Failed to retrieve RMIServer stub: javax.naming.ServiceUnavailableException [Root exception is java.rmi.ConnectException: Connection refused to host: 127.0.0.1; nested exception is: 
	java.net.ConnectException: Connection refused (Connection refused)]
```

Should be an easy fix though.

https://github.com/scylladb/java-driver/runs/7063421162?check_suite_focus=true

**Comment by fruch** [comment]
> This broke Java Driver’s integration tests.
> 
> ```
> Error:  Failures: 
> Error:    CCMBridgeTest.should_make_JMX_connection:46 » IO Failed to retrieve RMIServer stub: javax.naming.ServiceUnavailableException [Root exception is java.rmi.ConnectException: Connection refused to host: 127.0.0.1; nested exception is: 
> 	java.net.ConnectException: Connection refused (Connection refused)]
> ```
> 
> Should be an easy fix though.
> 
> https://github.com/scylladb/java-driver/runs/7063421162?check_suite_focus=true

It was hardcored 127.0.0.1 in the driver ?

**Comment by bhalevy** [comment]
> It was hardcored 127.0.0.1 in the driver ?

I guess in the driver integration tests.
I'd be surprised if the driver itself needs the JMX address.

**Comment by fruch** [comment]
@bhalevy I think the java driver is using it for stats (at least c-s is has a way to use the JMX port)

**Pull request merged**